### ISSUE TITLE
Tracking issue for pattern with by-move & by-ref bindings

### ISSUE THREAD

**Issue by Centril** [rust-lang/rust#68354]
This is the tracking issue for `#![feature(move_ref_pattern)]`, which allows patterns containing both by-ref and by-move bindings at the same time. For example: `(ref x, y, ref mut z)`.

## Implementation history

- On 2020-02-09, the implementation of the feature in https://github.com/rust-lang/rust/pull/68376 landed, thereby fixing https://github.com/rust-lang/rust/issues/45600. The PR was written by @Centril and reviewed by @matthewjasper.

**Comment by CryZe** [comment]
Does this also address moves in default match bindings like so?
```rust
let (x, y) = &(Vec::new(), 2);
```
where I want x to be a reference, but not y?

**Comment by Centril** [comment]
That's unrelated to this issue. (You can achieve `move y` with `mut y`, but this requires that `y` is bound by-copy as opposed to by-move, as you cannot move out of something held under a reference.)

**Comment by CryZe** [comment]
Oh wow, that is incredibly surprising that mut already somewhat achieves "moves" (yeah it's really bind by copy). Though this seems really dangerous as seen here:
```rust
let mut value = (2,);
let (mut access,) = &mut value;
access = 6;
```
Where it looks like the value would get modified, but instead we are actually modifying a fairly implicit copy. And I guess this doesn't work at all if we just want to bind by copy without actually modifying our copy.

**Comment by Centril** [comment]
- On 2020-02-09, the implementation of the feature in https://github.com/rust-lang/rust/pull/68376 landed. The PR was written by @Centril and reviewed by @matthewjasper.

As of now, we don't believe more tests would needed before we stabilize, but if you can think of some, please do add or propose them. What we're waiting now is probably for more time to pass.

**Comment by jjpe** [comment]
When could this feature potentially be stabilized, assuming no (major) issues are found?
I ask because I would like to optimize a derive-style proc-macro crate I'm writing. Currently the generated code contains a match on 2 by-ref bindings, but performing the optimization would result in one by-ref and one by-move binding in a single match.

Of course the response might be "just use another control flow structure, e.g. if-let", and in principle I could do that. In practice though, because it is code generation code, debugging is harder, and I'd prefer to keep the current code intact as much as possible. So much so that I'd rather wait for this feature to become stable if the amount of time it will take to stabilize is not too great.

**Comment by Centril** [comment]
So this feature hasn't seen any dogfooding inside the compiler itself, and I don't know what the testing situation in the ecosystem is like. However, the `src/test/ui` testsuite is extensive, so maybe that is enough. If @matthewjasper thinks this is good enough we could probably stabilize in 1.44.

**Comment by jjpe** [comment]
Meanwhile, we're a number of months further, and the world is different.
Has this feature been tested enough for inclusion yet? I'd really like to see this stabilized ASAP.

**Comment by jjpe** [comment]
Especially since the test suite is extensive, shouldn't this be stabilizable in time for the next stable release?

**Comment by matthewjasper** [comment]
Nominating for T-lang to discuss stabilization.

**Comment by nikomatsakis** [comment]
Discussing in the @rust-lang/lang meeting today:

We would love to see this stabilized. @matthewjasper if you or someone else wants to prepare a PR and stabilization report (showing interesting test cases and with a short description of the behavior to be stabilized), that would be great!

One thing that would be nice is seeing links to this feature being used "in the wild" (i.e., in the ecosystem), but maybe we don't know of any projects. (One could search for the feature gate...)

**Comment by joshtriplett** [comment]
It looks like [there are no uses of the feature gate outside of rustc and announcements/metadata](https://grep.app/search?q=move_ref_pattern&words=true). That shouldn't necessarily stop us going forward, but it seems worth noting.

This is the kind of feature, though, that people won't necessarily use nightly just to get.

**Comment by jjpe** [comment]
> One thing that would be nice is seeing links to this feature being used "in the wild" (i.e., in the ecosystem), but maybe we don't know of any projects. (One could search for the feature gate...)

I've consistently been asking for this feature precisely *because* I want to use it in the wild :)
Specifically, I've been developing the `deltoid` and `deltoid-derive` crates which, once I open source them, will make it a lot easier to do delta calculations on deeply nested data trees (a notable exclusion being types that use borrows, because you can't shove a newly created value into a borrow). 
The derive macro will use this feature once it is stabilized, but until then it necessarily operates with reduced performance. This is because in order to avoid this pattern (in my case, a tuple of an owned type and a borrow) I now need to borrow the owned type in the `match` expression, and clone from the destructured borrows that the `match` construct provides to me.
This feature being in stable will allow me to remove the artificial borrow-and-clone in the generated code.

As for how this pattern appears in my code in the first place, when possible I often `match` on a tuple of stuff rather than do nested `match`es, because rightward drift can be a real PITA.

**Comment by lzutao** [comment]
I should say that https://grep.app only scans about one million repositories on GitHub.
The more exact search should use GitHub search features: https://github.com/search?q=move_ref_pattern&type=Code
Through some of them just vendor Rust or Rust testsuite.

**Comment by Heliozoa** [comment]
FWIW I ran into this issue today when fetching the status of some process running on a server in a loop and trying to match a 'previous_status` variable and the new status to figure out if there's been any change. The function that caused the error looked something like this ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=06e56e69e0cbf5065a68fd70a176bf08))
```rust
#[derive(PartialEq, Eq)]
struct Finished {}

#[derive(PartialEq, Eq)]
struct Processing {
    status: ProcessStatus,
}

#[derive(PartialEq, Eq)]
enum ProcessStatus {
    One,
    Two,
    Three,
}

#[derive(PartialEq, Eq)]
enum Status {
    Finished(Finished),
    Processing(Processing),
}

fn check_result(_url: &str) -> Status {
    // fetch status from some server
    Status::Processing(Processing {
        status: ProcessStatus::One,
    })
}

fn wait_for_result(url: &str) -> Finished {
    let mut previous_status = None;
    loop {
        match check_result(url) {
            Status::Finished(f) => return f,
            Status::Processing(p) => {
                match (&mut previous_status, p.status) {
                    (None, status) => previous_status = Some(status), // first status
                    (Some(previous), status) if *previous == status => {} // no change, ignore
                    (Some(previous), status) => { // error!
                        // new status
                        *previous = status;
                    }
                }
            }
        }
    }
}
```
In my case the fix was easy enough, I realized I want to do the same processing that would go in the new status case with the first status as well, so I could just do
```rust

                    (Some(previous), status) if *previous == status => {} // no change, ignore
                    (_, status) => {
                        // new status
                        previous_status = Some(status);
                    }
```
Though in my case a workaround was trivial, this is what a use in the wild might've looked like if this was in stable today so I thought I'd post it. It felt like something that should work, so I'm glad to see that work is underway to stabilize it even if I'm not necessarily in a hurry to use it myself. Thanks!

**Comment by nikomatsakis** [comment]
Tagging with E-mentor -- this needs a stabilization PR, as described here:

https://rustc-dev-guide.rust-lang.org/stabilization_guide.html

The PR should include a brief report, but it doesn't have to go into details. The short version is:

* The older borrow checker prevents a single pattern from including *both* "by move" and "by ref" patterns (e.g., `Foo(x, ref y)`). This newer borrow checker can handle those soundly, but we still kept them feature gated.
* Rendering this feature gate unnecessary will highlight whatever tests we have, we should check to be sure that they seem representative when we see the PR.
* The lang-team [decided here](https://github.com/rust-lang/rust/issues/68354#issuecomment-654425636) that we'd like to see this stabilized already, so we can presumably jump straight to FCP.

**Comment by jjpe** [comment]
@niko I've taken a quick look at what's involved in writing such a stabilization PR.
I'm willing to do it, but I don't have the time at this precise moment for all of it. 

Is there a possibility to perhaps have 2 or more people share the honor and responsibility?
Or would the communication overhead act as noise and drown out the signal in such a case?

**Comment by nikomatsakis** [comment]
@jjpe seems like that would be fine to me, if somebody else wants to help out!

**Comment by Amjad50** [comment]
Hi @nikomatsakis @jjpe, I'm interested in helping with stabilizing this feature.

**Comment by jjpe** [comment]
While I love that there are now people willing to take this up, in the past month life has gotten much, much busier for me.
So even though I'd really like to do this, I cannot at this time.

**Comment by Amjad50** [comment]
I'll start working on stabilizing this feature

**Comment by Amjad50** [comment]
Process for stabilization:

- [x] Implementation (#68376)
- [x] [decision to stabilize](https://github.com/rust-lang/rust/issues/68354#issuecomment-654425636)
- [ ] Documentation (#76049)
- [ ] Stabilizing PR
- [ ] Stabilized

**Comment by Amjad50** [comment]
Does this feature removes Error `E0007`?

messge: `cannot bind by-move with sub-bindings`

Example:
```rust
```

**Comment by pacman82** [comment]
> One thing that would be nice is seeing links to this feature being used "in the wild" (i.e., in the ecosystem), but maybe we don't know of any projects. (One could search for the feature gate...)

I am using this feature in `odbc2parquet` without realizing it. Since the crate `parquet` is stable now, I'll probably try not to use it anymore until it is in the stable version of the compiler.

**Comment by memoryruins** [comment]
@pacman82 

> I'll probably try not to use it anymore until it is in the stable version of the compiler.

No worries, it is stable now!

The feature was stabilized in https://github.com/rust-lang/rust/pull/76119 and was released last month in stable rust 1.49 https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1490-2020-12-31 . It should be safe to close this tracking issue now ^^
